src/middlewared/middlewared/etc_files/local/ssh/config.py

```diff
@@ -1,36 +1,84 @@
-    ssh_config = middleware.call_sync('ssh.config')
-    for k in [
-        'ssh_host_key', 'ssh_host_key.pub', 'ssh_host_dsa_key', 'ssh_host_dsa_key.pub', 'ssh_host_dsa_key-cert.pub',
-        'ssh_host_ecdsa_key', 'ssh_host_ecdsa_key.pub', 'ssh_host_ecdsa_key-cert.pub', 'ssh_host_rsa_key',
-        'ssh_host_rsa_key.pub', 'ssh_host_rsa_key-cert.pub', 'ssh_host_ed25519_key', 'ssh_host_ed25519_key.pub',
-        'ssh_host_ed25519_key-cert.pub'
-    ]:
+DEFAULT_FILES = ['moduli', 'ssh_config', 'ssh_config.d', 'sshd_config', 'sshd_config.d']
+
+
+def generate_ssh_config(middleware, ssh_config, dirfd):
+    for k in SSH_KEYS:
+        must_remove = True
         s_key = re.sub(r'([.-])', '_', k).replace('ssh_', '', 1)
         if ssh_config[s_key]:
             decoded_key = base64.b64decode(ssh_config[s_key])
             if decoded_key:
-                with open(os.path.join(SSH_CONFIG_PATH, k), 'wb') as f:
+                mode = 0o644 if k.endswith('.pub') else 0o600
+
+                fd = os.open(k, os.O_WRONLY | os.O_CREAT, mode=mode, dir_fd=dirfd)
+                with open(fd, 'wb') as f:
+                    st = os.fstat(fd)
+                    if stat.S_ISREG(st.st_mode) == 0:
+                        middleware.logger.warning(
+                            "%s/%s: is not a regular file and will be removed. "
+                            "This may impact SSH access to the server.",
+                            SSH_CONFIG_PATH, k
+                        )
+                        with suppress(FileNotFoundError):
+                            os.remove(os.path.join(SSH_CONFIG_PATH, k))
+
+                    if stat.S_IMODE(st.st_mode) != mode:
+                        middleware.logger.debug(
+                            "%s/%s: file has unexpected permissions [%s]. "
+                            "Changing to new value [%s].",
+                            SSH_CONFIG_PATH, k, stat.S_IMODE(st.st_mode), mode
+                        )
+                        os.fchmod(f.fileno(), mode)
+
+                    if st.st_uid != 0 or st.st_gid != 0:
+                        middleware.logger.debug(
+                            "%s/%s: unexpected owner [%d:%d]. "
+                            "Changing to new value [0:0]. ",
+                            SSH_CONFIG_PATH, k, st.st_uid, st.st_gid
+                        )
+                        os.fchown(f.fileno(), 0, 0)
+
                     f.write(decoded_key)
+                    must_remove = False
+
+        if must_remove:
+            with suppress(FileNotFoundError):
+                os.remove(os.path.join(SSH_CONFIG_PATH, k))
 
-    for f in [
-        'ssh_host_key', 'ssh_host_dsa_key', 'ssh_host_ecdsa_key', 'ssh_host_rsa_key', 'ssh_host_ed25519_key'
-    ]:
-        if os.path.exists(os.path.join(SSH_CONFIG_PATH, f)):
-            os.chmod(os.path.join(SSH_CONFIG_PATH, f), 0o600)
+    expected_files = SSH_KEYS + DEFAULT_FILES
+    for f in os.listdir(dirfd):
```

Comment:
You'll want to test the code that I posted because `dirfd` is being passed to a context manager and I'm not sure if it closes it once it leaves scope so you might get `Bad File Descriptor` in the `def render` method. Not sure but just pointing it out before I forget.

Reply:
The context manager in this case doesn't close `dirfd`.